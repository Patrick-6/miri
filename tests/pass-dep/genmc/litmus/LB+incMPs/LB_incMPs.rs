//@compile-flags: -Zmiri-genmc

#![no_main]

#[path = "../../../../utils-dep/mod.rs"]
mod utils_dep;

use std::ffi::c_void;
use std::sync::atomic::{AtomicU64, Ordering};

static X: AtomicU64 = AtomicU64::new(0);
static Y: AtomicU64 = AtomicU64::new(0);
static W: AtomicU64 = AtomicU64::new(0);
static Z: AtomicU64 = AtomicU64::new(0);

use crate::utils_dep::*;

#[unsafe(no_mangle)]
fn miri_start(_argc: isize, _argv: *const *const u8) -> isize {
    // TODO GENMC: Hack (since Miri handles allocations lazily, and GenMC doesn't, we need to use them so they are `malloced` before any other thread uses them)
    X.store(0, Ordering::SeqCst);
    Y.store(0, Ordering::SeqCst);
    Z.store(0, Ordering::SeqCst);
    W.store(0, Ordering::SeqCst);

    let thread_order = [thread_1, thread_2, thread_3, thread_4];
    let _ids = unsafe { create_pthreads_no_params(thread_order) };

    0
}

extern "C" fn thread_1(_value: *mut c_void) -> *mut c_void {
    X.load(Ordering::Acquire);
    Z.fetch_add(1, Ordering::AcqRel);
    std::ptr::null_mut()
}

extern "C" fn thread_2(_value: *mut c_void) -> *mut c_void {
    Z.fetch_add(1, Ordering::AcqRel);
    Y.store(1, Ordering::Release);
    std::ptr::null_mut()
}

extern "C" fn thread_3(_value: *mut c_void) -> *mut c_void {
    Y.load(Ordering::Acquire);
    W.fetch_add(1, Ordering::AcqRel);
    std::ptr::null_mut()
}

extern "C" fn thread_4(_value: *mut c_void) -> *mut c_void {
    W.fetch_add(1, Ordering::AcqRel);
    X.store(1, Ordering::Release);
    std::ptr::null_mut()
}
