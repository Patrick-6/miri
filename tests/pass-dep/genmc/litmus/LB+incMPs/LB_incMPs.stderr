(GenMC Mode) Finished after 15 iterations.
